### PR TITLE
DS3231 v3

### PR DESCRIPTION
**Pull request by bhuvanchandra**
Changes since #167  :

    Fix typo in comment.

Changes since #166  :

    Use enums instead of defines.
    Add license string in the ds3231 driver.

### CODE REVIEW

**Comment by projectgus** [comment]
Hi @bhuvanchandra,

Thanks for these changes, looks good. The only other change, as mentioned in the other PR, is to change tabs to spaces (I missed this when looking at the original PR.) There are some coding guidelines here: https://github.com/SuperHouse/esp-open-rtos/#coding-style

Also, thanks for re-squashing your commits to keep the history clean & logical. You don't need to start a new PR each time, though. You can force push (git push -f) to the old PR branch and the PR will automatically update. I know it's considered a bad habit to force push, but for "your" PR branches it's OK. :)

Angus

**Comment by bhuvanchandra** [comment]
Hi @projectgus,

Fixed the tab indents and force pushed the changes.

--
Bhuvan

File: extras/ds3231/ds3231.h [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+enum {
+    DS3231_ALARM2_EVERY_MIN = 0,
+    DS3231_ALARM2_MATCH_MIN,
+    DS3231_ALARM2_MATCH_MINHOUR,
+    DS3231_ALARM2_MATCH_MINHOURDAY,
+    DS3231_ALARM2_MATCH_MINHOURDATE
+};
+
+bool ds3231_setTime(struct tm *time);
```
Comment (sheinz) [inline extras/ds3231/ds3231.h]: You have all those nice function comments in C file. Don't you think header is a better place for them? As a header is the first place people look when figuring out how to use a library.

Comment (bhuvanchandra) [inline extras/ds3231/ds3231.h]: Agreed, will update it.

**Comment by bhuvanchandra** [comment]
Hi @projectgus,

Also pushed the changes as per @sheinz comments.

--
Bhuvan

**Comment by ourairquality** [comment]
Started testing this and it looks ok so far. It seems fine to merge it.

One request is to change `ds3231_getTempInteger` to return a `uint16_t` and include all the 12 bits available rather than truncating to 8 bits. Then `ds3231_getTempFloat` could reuse the same code, or perhaps even be removed leaving it to the caller to convert to a float if they need to.

**Comment by ourairquality** [comment]
The casting '(signed)data[0]' in `ds3231_getTempFloat` might not be correct. Here's a suggestion that also returns all 12 bits for the integer version which could be used for the float version too:

```
bool ds3231_getTempInteger(int16_t *temp)
{
    uint8_t data[2];

    data[0] = DS3231_ADDR_TEMP;
    if (ds3231_send(data, 1) && ds3231_recv(data, 2)) {
        *temp = (int16_t)(int8_t)data[0] << 2 | data[1] >> 6;
        return true;
    }
    return false;
}
```

**Comment by sheinz** [comment]
I agree with @ourairquality that `(signed)data[0]` is not correct. But it will work nevertheless as `*temp` is `int8`.

With suggested by @ourairquality code I would suggest multiply the result by 25 and specify in the comment that function returns temperature in hundredth of a °C, for example 2525 is 25.25°C.

From the datasheet:
>Temperature is represented as a 10-bit code with a
resolution of 0.25°C and is accessible at location 11h and
12h. The temperature is encoded in two’s complement
format. The upper 8 bits, the integer portion, are at location
11h and the lower 2 bits, the fractional portion, are in
the upper nibble at location 12h. For example, 00011001
01b = +25.25°C. 

@bhuvanchandra Have you tested the negative temperature values? I'm not sure if the temperature in a register is a signed integer. It might just flip the sign bit without inverting the data bits. I haven't found information in the datasheet.

**Comment by bhuvanchandra** [comment]
@ourairquality and @sheinz, thanks! for your feedback and suggestions.

@ourairquality, agreed, will update it accordingly.

@sheinz i didn't test at negative temperature, i guess it will flip the sign bit for negative values. The only clue we have is that the temperature value(10-bit) is encoded in 2's complement.

How about this snippet:
```
bool ds3231_getTemperature(int16_t *temp)
{
    uint8_t data[2];

    data[0] = DS3231_ADDR_TEMP;

    if (ds3231_send(data, 1) && ds3231_recv(data, 2)) {
        /* Check the sign bit */
        if (data[0] >> 7) {
            *temp = (int16_t) ((((data[0] & 0x7F) << 2) | ((data[1] & 0XC0) >> 6)) - 1024) * 25;
            return true;
        }

        *temp = (int16_t) ((data[0] << 2) | ((data[1] & 0XC0) >> 6)) * 25;
        return true;
    }
    return false;
}
```

**Comment by ourairquality** [comment]
@sheinz Could an API that returns the binary fixed point representation generated by the device please be included as a minimum. Multiplying by 25 just frustrates most callers, causes premature overflow and loss of precision in subsequent integer computation. Personally I would not even include the `*25` and `float` API, but just request the binary fixed point API from which the other two variants can easily reuse.

@bhuvanchandra The encoding of the sign appears to be twos-complement (but I shall try to test) and the suggested code did work on some examples and is much simple. It's not necessary to mask out the low bits before the right shifts etc, lots of unnecessary code. I think the C standard guarantees that `int16_t` has a two's complement representation, but not certain, but it seems fine to me to assume all target devices use two's complement.

**Comment by ourairquality** [comment]
Just to be sure on the negative temperatures, put it in the freezer and with the above suggested code it made a clean transition below zero and back above zero which is consistent with it being two's complement and supports my code suggestion above.

**Comment by sheinz** [comment]
> Could an API that returns the binary fixed point representation generated by the device please be included as a minimum. Multiplying by 25 just frustrates most callers, causes premature overflow and loss of precision in subsequent integer computation. Personally I would not even include the *25 and float API, but just request the binary fixed point API from which the other two variants can easily reuse.

Then provide additional function, that might be named something like `ds3231_getRawTemperature` which would return pure data from the registers. 

In my opinion having those three function would cover all the users needs for the temperature:
   
    ds3231_getRawTemperature - will return pure register data(shifted)
    ds3231_getIntegerTemperature  -  will return only integer part of the temperature
    ds3231_getFloatTemperature
 
* names are just for reference, feel free to name it as you see appropriate

**Comment by projectgus** [comment]
Supplying those three options sounds ideal to me. I can see different use cases for each.

**Comment by bhuvanchandra** [comment]
@ourairquality yes, your code suggestion works nicely at negative temperatures also. I also tested it at my end today.
@sheinz, @projectgus, will add the three options.

--
Bhuvan

**Comment by bhuvanchandra** [comment]
Pushed the changes.

**Comment by kanflo** [comment]
Thanks @bhuvanchandra, this looks good so I'll go ahead and merge it.

**Pull request merged**